feat(aio): grouped search results

petebacondarwin opened the pull request:
The basic functionality is ready. The main feature missing is to hide the results when you click outside the results list. That is provided in #15096

petebacondarwin committed:
build(aio): generate the api-list.json file from the API docs

petebacondarwin committed:
fix(aio): `LocationService` search should handle corner cases

Adds tests and fixes corners cases for both `search()` and `setSearc()`
for things like empty queries and param keys that need encoding.

This commit refactors the `LocationService` to rely upon the `PlatformLocation`
rather than using `window.history` directly. This makes testing easier but also
makes the code simpler since `PlatformLocation` deals with platforms that
do not support history for us.

mary-poppins commented:
The angular.io preview for 0e278b5 is available [here][1].

[1]: https://pr15040-0e278b550cf5682b37b8d061f535eecb212d98ef.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 5d1de07 is available [here][1].

[1]: https://pr15040-5d1de0789a668f9c596116b5207071601212a7ec.ngbuilds.io/

gkalpak commented on aio/src/app/search/search-results/search-results.component.ts:
```diff
@@ -0,0 +1,73 @@
+import { Component, ChangeDetectionStrategy, EventEmitter, HostListener, Input, OnInit, Output } from '@angular/core';
```
`Input` not used.

petebacondarwin replied:
fixed

gkalpak commented on aio/src/app/search/search-results/search-results.component.ts:
```diff
@@ -0,0 +1,73 @@
+}
+
+/**
+ * A component to display the a list of search results
```
_the a_ --> _the_ or _a_ :smiley:

petebacondarwin replied:
fixed

gkalpak commented on aio/src/app/search/search-results/search-results.component.ts:
```diff
@@ -0,0 +1,73 @@
+  /**
+   * A mapping of the search results grouped into areas
+   */
+  searchAreas = new ReplaySubject<SearchArea[]>();
```
Do we need to replay anything but the last value?

petebacondarwin replied:
only one - good spot.

gkalpak commented on aio/src/app/search/search-results/search-results.component.spec.ts:
```diff
@@ -0,0 +1,109 @@
+    let selectedResult: SearchResult;
+    component.resultSelected.subscribe((result: SearchResult) => selectedResult = result);
+    const results = [
+      {path: 'news', title: 'News', type: 'marketing', keywords: '', titleWords: '' }
```
Inconsistent whitespace here and below: `{page...'' }` 😱 😱 😱

petebacondarwin replied:
fixed

gkalpak replied:
I don't think so 😛, but it shouldn't block this PR. I am pretty sure there are other places with inconsistent whitespace.

gkalpak commented on aio/src/app/search/search-box/search-box.component.ts:
```diff
@@ -0,0 +1,43 @@
+      // Ignore escape key
+      return;
+    }
+    this.locationService.setSearch('Full Text Search', { search: query });
```
I just noticed that the first argument to `setSearch` [is ignored][1]. Is it a bug? (Not related to this PR - just verifying.)

[1]: https://github.com/petebacondarwin/angular/blob/5d1de0789a668f9c596116b5207071601212a7ec/aio/src/app/shared/location.service.ts#L60

petebacondarwin replied:
It is a bug in that service, which is fixed in another of my PRs :-)

https://github.com/angular/angular/pull/15090/files#diff-d7053af3a910e9defc92631371cb06efR59

gkalpak commented on aio/src/app/search/search-box/search-box.component.spec.ts:
```diff
@@ -0,0 +1,43 @@
+    input.triggerEventHandler('focus', { target: { value: 'some query'}});
+    expect(search.search).toHaveBeenCalledWith('some query');
+  }));
+});
```
Missing a couple of tests:

- Calling `LocationService#setSearch()`.
- Ignoring the <kbd>ESC</kbd> key.
- Initializing based on URL query.

petebacondarwin replied:
added

gkalpak commented:
BTW, if the search is costly, we could avoid redoing it when focusing the search-box. We could cache the last result and return that if the query matches. But doesn't have to be part of this PR (if at all necessary in the first place).

petebacondarwin commented on aio/src/app/search/search-box/search-box.component.ts:
```diff
@@ -0,0 +1,43 @@
+export class SearchBoxComponent implements OnInit {
+
+  @ViewChild('searchBox') searchBox: ElementRef;
+  constructor(private searchService: SearchService, private locationService: LocationService) { }
```
done

petebacondarwin commented:
The search is done in a webworker so it is not costly from the user's perspective.
I agree that we could optimize it since we have the data available but I felt that it added unnecessary complexity for now practical gain.

petebacondarwin committed:
feat(aio): improve search results functionality

* Ensure that all indexed documents are displayed in the search results.
(Previously the guide documents were not appearing because we only showed
results that had a `name` property, rather than a `name` or `title`.)
* Group the results by their containing folder (e.g. api, guide, tutorial, etc).
* Hide the results when the user hits the ESC key.
* Hide the results when the user clicks on a search result

Closes #14852

petebacondarwin commented on aio/src/app/search/search-box/search-box.component.ts:
```diff
@@ -22,10 +22,12 @@ import { LocationService } from 'app/shared/location.service';
 
   ngOnInit() {
     const query = this.locationService.search()['search'];
+    console.log('here', query, this.locationService.search());
```
doh!

mary-poppins commented:
The angular.io preview for e937ad5 is available [here][1].

[1]: https://pr15040-e937ad5c3228cdb463f881514f6612dc8ed94c9a.ngbuilds.io/
